### PR TITLE
Progressive loading

### PR DESCRIPTION
**Pull request by prabhasp**
Up here for some scrutiny.

### CODE REVIEW

File: src/hatti/ona/stream.cljs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  "Calls callback on aggregated data that come via messages in channel.
+   Callback accepts two arguments: data (cljs list), and completed? (bool)
+  If there is a parser in message, resume it after a timeout."
+  [agg-data-channel callback]
+  (go
+   (while true
+     (let [{:keys [data completed? ]} (<! agg-data-channel)]
+       (when data (callback data completed?))
+       ;; Timeout so browser can do some rendering before parsing again.
+       (<! (timeout 10))
```
Comment (pld) [inline src/hatti/ona/stream.cljs]: this is millis I imagine? is there a reason this specific value was chosen? should we put it into a `def`?

is the reason for the timeout in the completed condition the same?

File: src/hatti/ona/stream.cljs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+       (when completed?
+         (<! (timeout 1000))
+         (close! agg-data-channel))))))
+
+(defn read-next-chunk!
+  "On each step of data, increment a counter *n*, and put data in a list *agg*.
+   After a exponentially increasing # of steps have been processes,
```
Comment (pld) [inline src/hatti/ona/stream.cljs]: `ed`

File: src/hatti/ona/stream.cljs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+   the aggregated data gets flushed into a channel, and parser is paused.
+   The channel reader should resume the parser after processing the data."
+  [data-chunk *n* *agg* channel]
```
Comment (pld) [inline src/hatti/ona/stream.cljs]: `atom`s don't need earmuffs, https://github.com/bbatsov/clojure-style-guide#atoms, https://github.com/bbatsov/clojure-style-guide#earmuffs-for-dynamic-vars

File: src/hatti/shared.cljs [hunk 4]
```diff
@@ -70,13 +67,22 @@
...
-    (when-not (zero? num_of_submissions)
-      (transact-app-state! app-state [:dataset-info :num_of_submissions] (fn [_] num_of_submissions)))))
+    (state/update-app-state! app-state [:status :total-records] total-records)))
+
+(defn add-to-app-data!
+  "Add to app data."
+  [app-state data & {:keys [completed?]}]
+  (let [old-data (:data @app-state)
+        new-data (concat old-data data)
+        total-count (count new-data)]
```
Comment (pld) [inline src/hatti/shared.cljs]: looks like you don't need this

**Comment by pld** [comment]
conflict and comments, but :+1:, looks good

**Comment by pld** [comment]
Hmm, it can't find the CLJS Oboe library

**Comment by prabhasp** [comment]
Oh right, I need to finish my pr with cljsjs. For now, you can pull
github.com/prabhasp/packages, go to the oboejs branch, navigate into the
oboe folder, and then run `boot package build-jar` locally.
Will get my pr going today.

**Comment by prabhasp** [comment]
`cljsjs/oboe` should now be merged / pushed to clojars, according to: https://github.com/cljsjs/packages/pull/224

**Comment by prabhasp** [comment]
Tests passing, pull request comments addressed, and rebased with master.

**Comment by pld** [comment]
Merged!!

**Comment by pld** [comment]
@prabhasp can you deploy to clojars and let's integrate it!

**Comment by prabhasp** [comment]
@pld deployed!

**Pull request merged**